### PR TITLE
Permit dynamically-sized "stack" allocations

### PR DESCRIPTION
**Pull request by abadams**
This is part 2 of the 1-2 punch to handle the cost of inner loop allocations going on the heap (and the consequent uselessness of GuardWithIf). The other part is #3276 

Currently if you try to `store_in(MemoryType::Stack)` an allocation that ends up having a dynamic size, you get a compile-time error. This is bad because it means the legality of a schedule depends on the quality of the simplifier. This PR gives these allocations stack-like behavior (while actually putting them on the heap). Stack-like behavior means:

1) They have a lifetime equal to the lifetime of the containing function
2) They are only allocated O(1) times per call to the containing function

This is giving a behavior to something that used to be a compile-time error, so it shouldn't affect any existing code. Beyond points 1 and 2, the implementation is opaque to the user and can be changed in the future if we want to do fancier allocation planning. The implementation is also opaque/not-overridable because this really needs to be force-inlined to be fast, so there's no opportunity for weak-linkage-based or function-pointer-based replacement of it.

The implementation is best explained by describing the other paths considered: At first I considered calling alloca, but this has to be done at function scope, and these dynamically sized allocations may well depend on loop variables in their size so they can't go there. llvm handles saving and restoring the stack at a deeper nesting that function scope poorly, and in practice stack size is often limited, so this seemed like a bad idea. Then I considered just mapping a whole bunch of address space to use as a side stack, but it would have to be thread-local, and you'd still have to check for overflows, so each allocation needs to have a comparison against the size followed by either a fast path or a slow path. It also needs per-OS support, and we need to convince people that using a ton of address space isn't as bad as using a ton of memory. Given that, a more conservative strategy is to use a deque as the side-stack, allocating new heap pages as necessary to back storage. But that has fragmentation issues, and is actually more complex than necessary: We know that at most one allocation is live per Func (per containing function - remember that a body of a parallel loop is a separate function), so it's even simpler to use a separate heap slot per buffer, with stack-equivalent lifetime, and sized for that buffer's high watermark.

The implementation therefore allocates a pointer and a size on the stack for each dynamic allocation with MemoryType::Stack. This is represented in code as a pseudostack_slot_t. It is initially (null, 0) and is allocated on first use. If the allocation is inside a loop and its size increases on a later loop iteration (unlikely - typically these things have a fixed size that is simply unknown until runtime) it is reallocated. It gets cleaned up on function exit (whether successfully or due to error). The lifetime is as if the buffer was on the stack, but the unknown dynamic size is handled with an indirection to the heap. Think of it as a unique_ptr local variable.

The PR includes a performance test that illustrates why this is a good thing. It places a small dynamically-sized allocation on the heap, in this pseudostack, and on the stack (using specialize to make the size an actual constant) and reports performance. Output on my laptop is:
```
Time using heap: 0.009903
Time using pseudostack: 0.001362
Time using stack: 0.001084
```

So it's way way faster than a heap allocation, and a bit slower than a true constant-sized stack allocation. The overhead over an actual stack allocation is due to the fact that you still need to check the dynamic size doesn't overflow, which requires some amount of integer math, and because the size and loop extent is dynamic, these things aren't going to get promoted into registers.

An alternative considered is to do a small-string-optimization-like thing and also allocate 32-64 bytes on the stack alongside the pointer and size to use in the case where the allocation is small enough to fit. This can be experimented with later if the cost of the single malloc call proves significant in some real pipeline.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Is this PR ready to review, or is more work needed?

**Comment by abadams** [comment]
I'm still considering it. It has some unintended consequences the way it currently stands.

**Comment by abadams** [comment]
ok, ready for review. There were two performance problems that I fixed:

If you mark something small and dynamic as Stack, BoundSmallAllocations tried to find a constant upper bound on the size, even if ludicrous, so that it could go on the stack. This is no longer a good idea.

Real stack allocations with non-overlapping lifetimes shared stack slots. Heap allocations with non-overlapping lifetimes also end up sharing memory because the allocator just hands it back to you. Pseudostack allocations were all live at the same time, increasing peak memory footprint by O(n) for producer-consumer chains of length n. I added logic to get them to share pseudostack slots, with reallocation as necessary. This gets you double-buffering in long producer-consumer chains, as you would want.

**Comment by steven-johnson** [comment]
I'll do an experimental pull-and-test into google

**Comment by steven-johnson** [comment]
For the record, this injected a few dozen new failures into our code:
- several halide_error_buffer_allocation_too_large
- at least one memory leak
- at least one segfault (@ 0x8)

**Comment by abadams** [comment]
Not sure how to proceed - valgrind on the tests reports clean. Any chance you could dig deeper into one of the new failures? A stack overflow seems like a possibility.

**Comment by steven-johnson** [comment]
I'll retry and dig a little deeper on some of the failures.

**Comment by abadams** [comment]
Thanks!

**Comment by steven-johnson** [comment]
Mind if I sync to master?

**Comment by abadams** [comment]
Go for it.

**Comment by steven-johnson** [comment]
Running under MSAN is giving me some use-of-inited-memory failures, but I haven't tracked down specifics yet

**Comment by steven-johnson** [comment]
Running under TSAN is giving me some heap-use-after-free failures: halide code is calling memset, and the halide code is running under a custom-but-well-tested ParallelFor() implementation

**Comment by steven-johnson** [comment]
For at least one of the buffer-too-large allocation failures:

    Error: Total allocation for buffer image_left_downsampled is 8589672450, which exceeds the maximum size of 2147483647

looking at debug codegen shows

    allocate image_left_downsampled$1[uint16 * -131071]

File: src/BoundSmallAllocations.cpp [hunk 1]
```diff
@@ -46,27 +47,32 @@ class BoundSmallAllocations : public IRMutator2 {
...
-            << "Only fixed-size allocations can be stored on the stack or in registers. "
-            << "Try storing on the heap instead.";
+            << "Only fixed-size allocations can be stored in registers. "
+            << "Try storing on the heap or stack instead.";
         user_assert(!in_thread_loop || bound.defined())
             << "Allocation " << op->name << " has a dynamic size. "
             << "Only fixed-size allocations are supported on the gpu. "
             << "Try storing into shared memory instead.";
+
+        if (bound.defined()) {
```
Comment (abadams) [inline src/BoundSmallAllocations.cpp]: yikes, I moved this simplify/cast before the pulling it out as an int64_t

**Comment by abadams** [comment]
review ping

**Comment by abadams** [comment]
It's not supposed to affect existing schedules, so no, I don't think so.

**Pull request merged**